pkg/rpc/server/server.go

```diff
@@ -1873,7 +1889,153 @@ func (s *Server) runScriptInVM(t trigger.Type, script []byte, contractScriptHash
+	s.sessionsLock.Lock()
+	sess := &session{
+		iterators: iterators,
+		finalize:  finalize,
+		till:      time.Now().Add(time.Second * time.Duration(s.config.SessionExpirationTime)),
+	}
+	s.sessions[sessionID] = sess
+	s.sessionsLock.Unlock()
+}
+
+func (s *Server) handleSessions() {
+	if !s.config.SessionEnabled {
+		return
+	}
+	s.log.Info("rpc-server iterator sessions are enabled")
+	var (
+		finalizationInterval = time.Duration(s.config.SessionExpirationTime) * time.Second
+		finalizationTicker   = time.NewTicker(finalizationInterval)
+	)
+
+	defer func() {
+		finalizationTicker.Stop()
+		s.sessionsLock.Lock()
+		defer s.sessionsLock.Unlock()
+		for _, session := range s.sessions {
+			// Concurrent iterator traversal may still be in process, thus need to protect iterators access.
+			session.iteratorsLock.Lock()
+			session.finalize()
+			session.iteratorsLock.Unlock()
+		}
+		s.sessions = nil
+	}()
+
+	for {
+		select {
+		case <-s.shutdown:
+			return
+		case <-finalizationTicker.C:
+			s.sessionsLock.Lock()
+			if len(s.sessions) == 0 {
+				s.sessionsLock.Unlock()
+				continue
+			}
+			now := time.Now()
+			staleSessions := s.sessions
+			// We can easily reuse existing map, but the ticker has tick period equal to the session expiration, so likely,
+			// the whole set of sessions will be removed from map at the subsequent tick, it's cheaper than using `delete`
+			// for all set of elements.
+			s.sessions = make(map[string]*session)
+			for id, sess := range staleSessions {
+				if sess.till.Before(now) {
+					sess.iteratorsLock.Lock()
+					sess.finalize()
+					sess.iteratorsLock.Unlock()
+				} else {
+					s.sessions[id] = sess
+				}
+			}
+			s.sessionsLock.Unlock()
+		}
+	}
+}
+
+func (s *Server) traverseIterator(reqParams request.Params) (interface{}, *response.Error) {
+	if !s.config.SessionEnabled {
+		return nil, response.NewInvalidRequestError("sessions are disabled")
+	}
+	sID, err := reqParams.Value(0).GetUUID()
+	if err != nil {
+		return nil, response.NewInvalidParamsError(fmt.Sprintf("invalid session ID: %s", err))
+	}
+	iID, err := reqParams.Value(1).GetUUID()
+	if err != nil {
+		return nil, response.NewInvalidParamsError(fmt.Sprintf("invalid iterator ID: %s", err))
+	}
+	count, err := reqParams.Value(2).GetInt()
+	if err != nil {
+		return nil, response.NewInvalidParamsError(fmt.Sprintf("invalid iterator items count: %s", err))
+	}
+	if err := checkInt32(count); err != nil {
+		return nil, response.NewInvalidParamsError("invalid iterator items count: not an int32")
+	}
+	if count > s.config.MaxIteratorResultItems {
+		return nil, response.NewInvalidParamsError(fmt.Sprintf("iterator items count is out of range (%d at max)", s.config.MaxIteratorResultItems))
+	}
+
+	s.sessionsLock.Lock()
+	session, ok := s.sessions[sID.String()]
+	if !ok {
+		s.sessionsLock.Unlock()
+		return []json.RawMessage{}, nil
+	}
+	session.till = time.Now().Add(time.Second * time.Duration(s.config.SessionExpirationTime))
```

Comment:
OK, timer recharge will be needed, but it can be done also.

Reply:
Fixed.